src/test.js

```diff
@@ -0,0 +1,105 @@
+import fs from 'fs';
+import { all, promisify, reject, resolve } from 'bluebird';
+import {
+    either,
+    evolve,
+    filter,
+    inc,
+    startsWith,
+    tryCatch
+} from 'ramda';
+import { compileES6 } from './compiler';
+import { getLocaleStrings } from './i18n';
+import { emitError, emitInfo, emitSuccess } from './input';
+import { inspect } from './module';
+import { compileSources } from './run';
+import { createVM, runAndGetAlerts } from './vm';
+
+const readFile = promisify(fs.readFile);
+
+/**
+ * Compiles the app files and returns an instance of a V8 object.
+ *
+ * @return {Promise<Object => Promise>}
+ */
+async function compileApp() {
+    const { name } = await readFile('package.json', 'utf-8')
+        | JSON.parse;
+    const [source, modules] = await compileSources();
+    const strings = await getLocaleStrings();
+    return context =>
+        runAndGetAlerts({ name, source }, context, strings, modules);
+}
+
+/**
+ * Compiles the test source.
+ *
+ * @return {Promise}
+ */
+async function compileTest() {
+    const source = await (readFile('test/index.js', 'utf-8')
+        .catchThrow(new Error('no tests provided')))
+        | compileES6;
+    const localTestModules = inspect(source).modules
+        | filter(either(startsWith('./'), startsWith('../')));
+
+    if (localTestModules.length > 0) {
+        return reject(new Error('only external modules can be required in testsuite. Found '
+            + localTestModules.join(', ')));
+    }
+
+    return source;
+}
+
+/**
+ * Executes the tests and returns a promise with computed results.
+ * @param {Object => Promise} runWithContext - Pre-compiled application
+ * @param {(String, Function)[]} tests - The test cases
+ * @return {Promise}
+ */
+async function runTests(runWithContext, tests) {
+    if (tests.length === 0) {
+        return emitInfo('no tests to run');
+    }
+
+    const loop = async ([test, ...rest], report = { passing: 0, failing: 0 }) => {
+        if (!test) {
+            const message = `${report.passing} passing, ${report.failing} failing`;
+            if (report.failing > 0) {
+                return reject(new Error(message));
+            }
+
+            return emitSuccess(message);
+        }
+
+        const [description, implementation] = test;
+        return runWithContext
+            | tryCatch(implementation & resolve, reject)
+            | (future => future
+                .then(~emitSuccess(description)
+                    .return(report | evolve({ passing: inc })))
+                .catch(err =>
+                    emitError(`${description}:\n${err.stack}\n`)
+                        .return(report | evolve({ failing: inc })))
+                .then(loop(rest, _)));
+    };
+
```

Comment:
Eu sei que tu vai me xingar, mas ta aí:

```javascript
let [passing, failing] = [0, 0]

for (let test in tests) {
    const [description, implementation] = test
    try {
        await runWithContext(implementation)
        emitSuccess(description)
        passing++
    } catch (e) {
        emitError(description)
        failing++
    }
}

const message = `${passing} passing, ${failing} failing`;
if (failing > 0) {
    return reject(new Error(message));
}
return emitSuccess(message);
```

Achei mais elegante que a solução recursiva, me julgue :angel:

Reply:
SHAME ON YOU!